### PR TITLE
Added `if_disabled` attribute for `generate` block

### LINKED ISSUE #3096: When `disable: true` on `generate` block after it has been enabled, cached file is not removed
**Describe the bug**
When a block is generated, it's not removed when the `generate` block is disabled. Manual cache invalidation is needed.

**To Reproduce**
First run this Terragrunt config

```hcl
generate "test" {
  disable = false
  path = "test.tf"
  if_exists = "overwrite"
  contents = <
output "test" {
  value = "test"
}
EOT
}
 
terraform {
  source = "./tf/"
}
```

terragrunt plan

```
marco@test$ terragrunt plan
Initializing the backend...
 
Initializing provider plugins...
 
Terraform has been successfully initialized!
 
...
 
Changes to Outputs:
  + test = "test"
 
...
```

Set to disable

```hcl
generate "test" {
  disable   = true
  path      = "test.tf"
  if_exists = "overwrite"
  contents  = <<EOT
output "test" {
  value = "test"
}
EOT
}

terraform {
  source = "./tf/"
}
```

Nothing changed:

```
marco@test$ terragrunt plan
Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

...

Changes to Outputs:
  + test = "test"

...
```

Flushing `.terragrunt-cache` cache, and with it the generated [test.tf](http://test.tf)

```
marco@test$ rm -rf ./.terragrunt-cache
marco@test$ terragrunt plan
INFO[0000] Downloading Terraform configurations from file:///home/marco/notes/testtf/tf into /home/marco/notes/testtf/.terragrunt-cache/wfIPpErBuUByitLEL5eelRC_g0s/vBTqYIOukLa735JrmTnCK9zPHsY 

Initializing the backend...

Initializing provider plugins...

Terraform has been successfully initialized!

No changes. Your infrastructure matches the configuration.

...
```

**Expected behavior**
Removal of the generated file if it has the signature.

**Nice to have**

- [X] Terminal output
- [ ] Screenshots

**Versions**

* Terragrunt version: v0.55.13
* Terraform version:
* Environment details (Ubuntu 20.04, Windows 10, etc.): WSL ubuntu 22.04

**Additional context**
[The disable feature PR](https://github.com/gruntwork-io/terragrunt/pull/2497)

**Workaround**
For now you can use the following to clear the contents of the file and leave an empty file when disabled:

```hcl
generate "test" {
  #disable = true
  path = "test.tf"
  if_exists = "overwrite"
contents = !local.enabled_test_tf ? "" : <
  value = "test"
}
EOT
}
 
terraform {
  source = "./tf/"
}
```

### PR DESCRIPTION
**Pull request by levkohimins**
## Description

Fixes #3096.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [ ] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added `if_disabled` attribute for `generate` block

### CODE REVIEW

File: codegen/generate.go [hunk 1]
```diff
@@ -65,21 +81,32 @@ type GenerateConfig struct {
...
-
 	targetFileExists := util.FileExists(targetPath)
+
+	// If this GenerateConfig is disabled then skip further processing.
+	if config.Disable {
+		terragruntOptions.Logger.Debugf("Skipping generating file at %s because it is disabled", config.Path)
+
+		if targetFileExists {
+			if shouldRemove, err := shouldRemoveWithFileExists(terragruntOptions, targetPath, config.IfDisabled); err != nil {
+				return err
```
Comment (denis256) [inline codegen/generate.go:99]: Make sense to wrap in `errors.WithStackTrace(err)`?  same for other places where err is returned

Comment (levkohimins) [reply codegen/generate.go:99]: > Make sense to wrap in `errors.WithStackTrace(err)`? same for other places where err is returned

No, we only have to wrap the call from the 3rd party library, `shouldRemoveWithFileExists` already has our error wrappers, if we wrap here the stack trace will start here and not from inside `mustRemoveWithFileExists`.

File: codegen/generate.go [hunk 2]
```diff
@@ -65,21 +81,32 @@ type GenerateConfig struct {
...
+			} else if shouldRemove {
```
[the remark this answers is not in the document]
Comment (levkohimins) [reply codegen/generate.go:100]: The scope of `shouldRemove` is only in the `if` clause. There are two ways to write this expression:

**short**

```go
			if shouldRemove, err := shouldRemoveWithFileExists(terragruntOptions, targetPath, config.IfDisabled); err != nil {
				return err
			} else if shouldRemove {
				if err := os.Remove(targetPath); err != nil {
					return errors.WithStackTrace(err)
				}
			}
```

**long** 

```go
			shouldRemove, err := shouldRemoveWithFileExists(terragruntOptions, targetPath, config.IfDisabled)
			if err != nil {
				return err
			}
			if shouldRemove {
				if err := os.Remove(targetPath); err != nil {
					return errors.WithStackTrace(err)
				}
			}
```

I deliberately wrote in a short form, but if you would like I can change it.

**Review by denis256 — APPROVED** [review]

**Comment by levkohimins** [comment]
Resolved in [v0.57.5](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.57.5) release.

**Pull request merged**